scorch conjuncts match phrase

steveyen opened the pull request:
While looking through some recent testrunner RQG runs, saw another case where there was a search result discrepancy.  This WIP change replicates this in the scorch-vs-upsidedown unit test.

From http://qa.sc.couchbase.com/job/test_suite_executor/39085/consoleText...

> [2017-12-18 22:07:02,054] - [task:1312] INFO - FTS hits for query: {"conjuncts": [{"field": "manages.reports", "match": "Balandria Cynara"}, {"field": "manages.reports", "match_phrase": "Maura"}]} is 5 (took 5.586473ms)
> [2017-12-18 22:07:02,060] - [task:1322] INFO - ES hits for query: {"query": {"bool": {"must": [{"match": {"manages.reports": "Balandria Cynara"}}, {"match_phrase": {"manages.reports": "Maura"}}]}}} on es_index is 7 (took 2ms)
> [2017-12-18 22:07:02,060] - [task:1327] ERROR - FAIL: FTS hits: 5, while ES hits: 7
>

steveyen commented:
Just putting the test case up here... but still marking this "Work In Progress" as there isn't an actual fix yet.

mschoch commented:
This test harness is great @steveyen.  I spent some time trying to look into this.  The index data for the missing hit seems OK.  So, I suspect the issue is on the term-field-reader or searcher side of things.  I tried simplifying the queries to remove some of the layers, and this query still seems to fail:

```
	`{
      "about": "conjuncts-match-phrase-1-1 inspired by testrunner RQG issue -- see: MB-27291",
      "query": {
        "conjuncts": [
          {"field": "body", "term": "{{bodyWord 0}}"},
          {"field": "body", "terms": ["{{bodyWord 1}}", "{{bodyWord 1}}"]}
        ]
      }
     }`,
```
It switches from match query to term query (removing disjunction searcher of terms) and switches from match phrase to regular phrase (again possibly removing extra layers of disjunction searchers).

mschoch commented:
Seems like it might be related to the behavior when a conjunction searcher is nested inside another conjunction searcher.  Perhaps another case where we do something unexpected to the term-field-reader.

Need to get some sleep now...

steveyen commented:
Indeed - I don't have a good handle on this either.  In the latest push/refresh to this PR just now, I tried sprinkling some more protective if-checks in various searcher.Advance() methods to try to prevent backwards seeking... but these didn't work (more details below) -- so I must have missed some case.

Additionally, this whack-a-mole approach, though, is starting to have a gut feel of a bit of a wrong path for me, as I don't feel I have a good systemic feel for it.

Now I'm starting to wonder whether a better (more comprehensive) track might be somehow to improve scorch's index-snapshot-termFieldReader to be able to support both generic forwards & backwards Advance() directions.

--------------
For kicks, I tried adding some checking code into store/boltdb's iterator Seek() method, to panic whenever it receives an invocation for a seek to a previous key.  Even with all the little checks that I'd added to try to avoid backwards Advance() calls at the higher searcher levels, this panic in store/boltdb still fires.

```
+++ b/index/store/boltdb/iterator.go
@@ -55,7 +56,21 @@ func (i *Iterator) Seek(k []byte) {
                        return
                }
        }
+
+ prevKey := i.key
+
+ backwards := false
+ if i.key != nil && bytes.Compare(k, i.key) < 0 {
+         backwards = true
+ }
+
        i.key, i.val = i.cursor.Seek(k)
+
+ if backwards && !bytes.Equal(prevKey, i.key) {
+         fmt.Printf("bolt seekToKey (%v) < prevKey (%v), now at: %v\n", k, prevKey, i.key)
+         panic("saw an actual boltdb backwards seek")
+ }
+
        i.updateValid()
 }
```

Finally, btw, when the panic() does fire, I do see those nested conjunction searchers that you were suspecting...

```
bolt seekToKey ([116 1 0 52 255 57 57 54]) < prevKey ([116 1 0 53 255 48]), now at: [116 1 0 52 255 57 57 54]

panic: saw an actual boltdb backwards seek [recovered]
	panic: saw an actual boltdb backwards seek

goroutine 9 [running]:
testing.tRunner.func1(0xc4201de0f0)
	/usr/local/Cellar/go/1.9/libexec/src/testing/testing.go:711 +0x2d2
panic(0x156a2a0, 0x16a5da0)
	/usr/local/Cellar/go/1.9/libexec/src/runtime/panic.go:491 +0x283
github.com/blevesearch/bleve/index/store/boltdb.(*Iterator).Seek(0xc420451180, 0xc42044cde8, 0x8, 0x8)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/index/store/boltdb/iterator.go:71 +0x57c
github.com/blevesearch/bleve/index/upsidedown.(*UpsideDownCouchTermFieldReader).Advance(0xc4208e9bc0, 0xc42044cd58, 0x3, 0x8, 0xc4202f9028, 0xc4202f8f28, 0xc420267b10, 0xc4205925e8)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/index/upsidedown/reader.go:132 +0x184
github.com/blevesearch/bleve/search/searcher.(*TermSearcher).Advance(0xc4202f9000, 0xc420267b20, 0xc42044cd58, 0x3, 0x8, 0x1, 0x1, 0xc420591f00)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_term.go:96 +0xc6
github.com/blevesearch/bleve/search/searcher.(*ConjunctionSearcher).advanceChild(0xc420418240, 0xc420267b20, 0x0, 0xc42044cd58, 0x3, 0x8, 0xc42044cd58, 0x3)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_conjunction.go:202 +0x196
github.com/blevesearch/bleve/search/searcher.(*ConjunctionSearcher).Advance(0xc420418240, 0xc420267b20, 0xc42044cd58, 0x3, 0x8, 0x8, 0x0, 0xc4203cc848)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_conjunction.go:190 +0x10b
github.com/blevesearch/bleve/search/searcher.(*PhraseSearcher).Advance(0xc4205233b0, 0xc420267b20, 0xc42044cd58, 0x3, 0x8, 0xc420591f00, 0x0, 0x0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_phrase.go:323 +0x9d
github.com/blevesearch/bleve/search/searcher.(*ConjunctionSearcher).advanceChild(0xc420418300, 0xc420267b20, 0x1, 0xc42044cd58, 0x3, 0x8, 0x1, 0x0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_conjunction.go:202 +0x196
github.com/blevesearch/bleve/search/searcher.(*ConjunctionSearcher).Next(0xc420418300, 0xc420267b20, 0x1930120, 0xc420671e40, 0xc420593088)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/searcher/search_conjunction.go:149 +0x132
github.com/blevesearch/bleve/search/collector.(*TopNCollector).Collect(0xc42061b110, 0x192a760, 0xc4200140e0, 0x192e620, 0xc420418300, 0x1930120, 0xc420671e40, 0xc420418300, 0x0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/search/collector/topn.go:138 +0x185
github.com/blevesearch/bleve.(*indexImpl).SearchInContext(0xc42022c000, 0x192a760, 0xc4200140e0, 0xc4201fc620, 0x0, 0x0, 0x0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/index_impl.go:432 +0xeda
github.com/blevesearch/bleve.(*indexImpl).Search(0xc42022c000, 0xc4201fc620, 0xc420523360, 0x0, 0x0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/index_impl.go:363 +0x4d
github.com/blevesearch/bleve/test.testVersusSearches(0xc4201ca230, 0x19a8a60, 0xd, 0xd, 0x1930a60, 0xc4201ca2a0, 0x1930a60, 0xc42022c000)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/test/versus_test.go:300 +0x68f
github.com/blevesearch/bleve/test.(*VersusTest).run(0xc4201ca230, 0x16405eb, 0x6, 0x16404d1, 0x6, 0x164229d, 0xb, 0x16404d1, 0x6, 0x165e228, ...)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/test/versus_test.go:434 +0x554
github.com/blevesearch/bleve/test.TestScorchVersusUpsideDownBoltAll(0xc4201de0f0)
	/Users/steveyen/go/src/github.com/blevesearch/bleve/test/versus_test.go:52 +0xf1
testing.tRunner(0xc4201de0f0, 0x165e200)
	/usr/local/Cellar/go/1.9/libexec/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.9/libexec/src/testing/testing.go:789 +0x2de
```

mschoch commented:
Thanks Steve, ya it at least feels like we are close to pinpointing the cause.  And I agree, might be better for me to support backwards Advance() and treat avoiding going backwards as a longer-term improvement.

steveyen commented:
Just pushed up a brute force attempt at supporting backwards Advance in the scorch term-field-reader... by just restarting the posting iterators the way at the beginning.  And, doing that functionally works (tests no longer failing) so it confirms some theories in my head about what's going on (in which I still feel I don't have the full picture), but it'd be horribly slow.

mschoch commented:
Cool, I didn't look at it yet, but it seems like it should be straightforward to start over at the beginning of the correct chunk for the new ID, and not necessarily all the way at the beginning.

mschoch commented:
OK, I see now that there are 2 improvements.  One, based on the advanced doc number, we can start directly on the correct segment.  Then, if we introduce a new Advance() method onto the PostingsIterator that will help as well.  Dumb implementations (in memory) just do Next, Next, Next, but chunked impls can skip over data.

I think we should merge this as is, and then I'll look at introducing these other fixes without breaking anything...

mschoch commented:
Hey Steve, so looking at this closer, do you know how you triggered that panic in bolt which shows actually seeking backwards?  After merging these changes, I made some slight tweaks:

1.  Instead of your >= 0 check, I kept my previous check for == 0 as a separate case, as this case can be handled with much less overhead (just return the same thing as last time, which I had already added code to remember).
2.  In the > 0 case, I added a panic which prints the keys.

At first, it did still panic, but it was an odd case where i.currPosting and i.currID weren't correctly set to nil in the case where we got to the end of the iterator.  Once I addressed this, the code no longer panics for me

But, the output you showed was pretty clearly going back to a previous key.  Is there any chance that was before you added all of the if/comparison checks?

steveyen commented:
> But, the output you showed was pretty clearly going back to a previous key. Is there any chance that was before you added all of the if/comparison checks?

Hi Marty,
Yah, that panic() for the backwards bolt seek was done after I'd added all those extra if/comparison checks.
-- steve


________________________________
From: Marty Schoch [notifications@github.com]
Sent: Sunday, December 24, 2017 4:32 AM
To: blevesearch/bleve
Cc: Steve Yen; Mention
Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)


Hey Steve, so looking at this closer, do you know how you triggered that panic in bolt which shows actually seeking backwards? After merging these changes, I made some slight tweaks:

  1.  Instead of your >= 0 check, I kept my previous check for == 0 as a separate case, as this case can be handled with much less overhead (just return the same thing as last time, which I had already added code to remember).
  2.  In the > 0 case, I added a panic which prints the keys.

At first, it did still panic, but it was an odd case where i.currPosting and i.currID weren't correctly set to nil in the case where we got to the end of the iterator. Once I addressed this, the code no longer panics for me

But, the output you showed was pretty clearly going back to a previous key. Is there any chance that was before you added all of the if/comparison checks?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/blevesearch/bleve/pull/692#issuecomment-353781797>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAB3sAFoKbxKmbFAxV8kWASuUk_mmtDPks5tDkRQgaJpZM4RLTy3>.

mschoch commented:
Do you know which test case triggers it?

On Dec 24, 2017 4:58 PM, "Steve Yen" <notifications@github.com> wrote:

> > But, the output you showed was pretty clearly going back to a previous
> key. Is there any chance that was before you added all of the if/comparison
> checks?
>
> Hi Marty,
> Yah, that panic() for the backwards bolt seek was done after I'd added all
> those extra if/comparison checks.
> -- steve
>
>
> ________________________________
> From: Marty Schoch [notifications@github.com]
> Sent: Sunday, December 24, 2017 4:32 AM
> To: blevesearch/bleve
> Cc: Steve Yen; Mention
> Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
>
>
> Hey Steve, so looking at this closer, do you know how you triggered that
> panic in bolt which shows actually seeking backwards? After merging these
> changes, I made some slight tweaks:
>
> 1. Instead of your >= 0 check, I kept my previous check for == 0 as a
> separate case, as this case can be handled with much less overhead (just
> return the same thing as last time, which I had already added code to
> remember).
> 2. In the > 0 case, I added a panic which prints the keys.
>
> At first, it did still panic, but it was an odd case where i.currPosting
> and i.currID weren't correctly set to nil in the case where we got to the
> end of the iterator. Once I addressed this, the code no longer panics for me
>
> But, the output you showed was pretty clearly going back to a previous
> key. Is there any chance that was before you added all of the if/comparison
> checks?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> blevesearch/bleve/pull/692#issuecomment-353781797>, or mute the thread<
> https://github.com/notifications/unsubscribe-auth/
> AAB3sAFoKbxKmbFAxV8kWASuUk_mmtDPks5tDkRQgaJpZM4RLTy3>.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/blevesearch/bleve/pull/692#issuecomment-353791596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASMny71lOJyJsCAibhH0i682GbzCiT1ks5tDnSmgaJpZM4RLTy3>
> .
>

steveyen commented:
> Do you know which test case triggers it?

Hi Marty,

I was definitely in the world of those "versus" tests at the time, so I'd either be running "go test -run Versus ./test" or I'd be into one of the smaller, specific versus tests like...

  TestScorchVersusUpsideDownBoltSmallMNSAM
  TestScorchVersusUpsideDownBoltSmallCMP11

-- steve

________________________________
From: Marty Schoch [notifications@github.com]
Sent: Sunday, December 24, 2017 8:31 AM
To: blevesearch/bleve
Cc: Steve Yen; Mention
Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)

Do you know which test case triggers it?

On Dec 24, 2017 4:58 PM, "Steve Yen" <notifications@github.com> wrote:

> > But, the output you showed was pretty clearly going back to a previous
> key. Is there any chance that was before you added all of the if/comparison
> checks?
>
> Hi Marty,
> Yah, that panic() for the backwards bolt seek was done after I'd added all
> those extra if/comparison checks.
> -- steve
>
>
> ________________________________
> From: Marty Schoch [notifications@github.com]
> Sent: Sunday, December 24, 2017 4:32 AM
> To: blevesearch/bleve
> Cc: Steve Yen; Mention
> Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
>
>
> Hey Steve, so looking at this closer, do you know how you triggered that
> panic in bolt which shows actually seeking backwards? After merging these
> changes, I made some slight tweaks:
>
> 1. Instead of your >= 0 check, I kept my previous check for == 0 as a
> separate case, as this case can be handled with much less overhead (just
> return the same thing as last time, which I had already added code to
> remember).
> 2. In the > 0 case, I added a panic which prints the keys.
>
> At first, it did still panic, but it was an odd case where i.currPosting
> and i.currID weren't correctly set to nil in the case where we got to the
> end of the iterator. Once I addressed this, the code no longer panics for me
>
> But, the output you showed was pretty clearly going back to a previous
> key. Is there any chance that was before you added all of the if/comparison
> checks?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> blevesearch/bleve/pull/692#issuecomment-353781797>, or mute the thread<
> https://github.com/notifications/unsubscribe-auth/
> AAB3sAFoKbxKmbFAxV8kWASuUk_mmtDPks5tDkRQgaJpZM4RLTy3>.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/blevesearch/bleve/pull/692#issuecomment-353791596>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASMny71lOJyJsCAibhH0i682GbzCiT1ks5tDnSmgaJpZM4RLTy3>
> .
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/blevesearch/bleve/pull/692#issuecomment-353793152>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAB3sLvvZkRppkmkcoyT_MA1YPlPkC9Qks5tDnx2gaJpZM4RLTy3>.

mschoch commented:
Ok I'll keep trying, must have run that a thousand times...

On Dec 24, 2017 5:48 PM, "Steve Yen" <notifications@github.com> wrote:

> > Do you know which test case triggers it?
>
> Hi Marty,
>
> I was definitely in the world of those "versus" tests at the time, so I'd
> either be running "go test -run Versus ./test" or I'd be into one of the
> smaller, specific versus tests like...
>
> TestScorchVersusUpsideDownBoltSmallMNSAM
> TestScorchVersusUpsideDownBoltSmallCMP11
>
> -- steve
>
> ________________________________
> From: Marty Schoch [notifications@github.com]
> Sent: Sunday, December 24, 2017 8:31 AM
> To: blevesearch/bleve
> Cc: Steve Yen; Mention
> Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
>
> Do you know which test case triggers it?
>
> On Dec 24, 2017 4:58 PM, "Steve Yen" <notifications@github.com> wrote:
>
> > > But, the output you showed was pretty clearly going back to a previous
> > key. Is there any chance that was before you added all of the
> if/comparison
> > checks?
> >
> > Hi Marty,
> > Yah, that panic() for the backwards bolt seek was done after I'd added
> all
> > those extra if/comparison checks.
> > -- steve
> >
> >
> > ________________________________
> > From: Marty Schoch [notifications@github.com]
> > Sent: Sunday, December 24, 2017 4:32 AM
> > To: blevesearch/bleve
> > Cc: Steve Yen; Mention
> > Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
> >
> >
> > Hey Steve, so looking at this closer, do you know how you triggered that
> > panic in bolt which shows actually seeking backwards? After merging these
> > changes, I made some slight tweaks:
> >
> > 1. Instead of your >= 0 check, I kept my previous check for == 0 as a
> > separate case, as this case can be handled with much less overhead (just
> > return the same thing as last time, which I had already added code to
> > remember).
> > 2. In the > 0 case, I added a panic which prints the keys.
> >
> > At first, it did still panic, but it was an odd case where i.currPosting
> > and i.currID weren't correctly set to nil in the case where we got to the
> > end of the iterator. Once I addressed this, the code no longer panics
> for me
> >
> > But, the output you showed was pretty clearly going back to a previous
> > key. Is there any chance that was before you added all of the
> if/comparison
> > checks?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub<https://github.com/
> > blevesearch/bleve/pull/692#issuecomment-353781797>, or mute the thread<
> > https://github.com/notifications/unsubscribe-auth/
> > AAB3sAFoKbxKmbFAxV8kWASuUk_mmtDPks5tDkRQgaJpZM4RLTy3>.
> >
> > —
> > You are receiving this because you modified the open/close state.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/blevesearch/bleve/pull/692#issuecomment-353791596>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AASMny71lOJyJsCAibhH0i682GbzCiT1ks5tDnSmgaJpZM4RLTy3>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> blevesearch/bleve/pull/692#issuecomment-353793152>, or mute the thread<
> https://github.com/notifications/unsubscribe-auth/AAB3sLvvZkRppkmkcoyT_
> MA1YPlPkC9Qks5tDnx2gaJpZM4RLTy3>.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/blevesearch/bleve/pull/692#issuecomment-353793860>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASMnwxBmyvUzTCwJ1RygWMU_9HDkRZ0ks5tDoBqgaJpZM4RLTy3>
> .
>

steveyen commented:
That's strange.  It fails on my machine right away.  One thing is the sample data is math.Rand() generated, but there is rand seeding to zero.  Weird.

-- steve


________________________________
From: Marty Schoch [notifications@github.com]
Sent: Sunday, December 24, 2017 9:06 AM
To: blevesearch/bleve
Cc: Steve Yen; Mention
Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)

Ok I'll keep trying, must have run that a thousand times...

On Dec 24, 2017 5:48 PM, "Steve Yen" <notifications@github.com> wrote:

> > Do you know which test case triggers it?
>
> Hi Marty,
>
> I was definitely in the world of those "versus" tests at the time, so I'd
> either be running "go test -run Versus ./test" or I'd be into one of the
> smaller, specific versus tests like...
>
> TestScorchVersusUpsideDownBoltSmallMNSAM
> TestScorchVersusUpsideDownBoltSmallCMP11
>
> -- steve
>
> ________________________________
> From: Marty Schoch [notifications@github.com]
> Sent: Sunday, December 24, 2017 8:31 AM
> To: blevesearch/bleve
> Cc: Steve Yen; Mention
> Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
>
> Do you know which test case triggers it?
>
> On Dec 24, 2017 4:58 PM, "Steve Yen" <notifications@github.com> wrote:
>
> > > But, the output you showed was pretty clearly going back to a previous
> > key. Is there any chance that was before you added all of the
> if/comparison
> > checks?
> >
> > Hi Marty,
> > Yah, that panic() for the backwards bolt seek was done after I'd added
> all
> > those extra if/comparison checks.
> > -- steve
> >
> >
> > ________________________________
> > From: Marty Schoch [notifications@github.com]
> > Sent: Sunday, December 24, 2017 4:32 AM
> > To: blevesearch/bleve
> > Cc: Steve Yen; Mention
> > Subject: Re: [blevesearch/bleve] scorch conjuncts match phrase (#692)
> >
> >
> > Hey Steve, so looking at this closer, do you know how you triggered that
> > panic in bolt which shows actually seeking backwards? After merging these
> > changes, I made some slight tweaks:
> >
> > 1. Instead of your >= 0 check, I kept my previous check for == 0 as a
> > separate case, as this case can be handled with much less overhead (just
> > return the same thing as last time, which I had already added code to
> > remember).
> > 2. In the > 0 case, I added a panic which prints the keys.
> >
> > At first, it did still panic, but it was an odd case where i.currPosting
> > and i.currID weren't correctly set to nil in the case where we got to the
> > end of the iterator. Once I addressed this, the code no longer panics
> for me
> >
> > But, the output you showed was pretty clearly going back to a previous
> > key. Is there any chance that was before you added all of the
> if/comparison
> > checks?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub<https://github.com/
> > blevesearch/bleve/pull/692#issuecomment-353781797>, or mute the thread<
> > https://github.com/notifications/unsubscribe-auth/
> > AAB3sAFoKbxKmbFAxV8kWASuUk_mmtDPks5tDkRQgaJpZM4RLTy3>.
> >
> > —
> > You are receiving this because you modified the open/close state.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/blevesearch/bleve/pull/692#issuecomment-353791596>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AASMny71lOJyJsCAibhH0i682GbzCiT1ks5tDnSmgaJpZM4RLTy3>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<https://github.com/
> blevesearch/bleve/pull/692#issuecomment-353793152>, or mute the thread<
> https://github.com/notifications/unsubscribe-auth/AAB3sLvvZkRppkmkcoyT_
> MA1YPlPkC9Qks5tDnx2gaJpZM4RLTy3>.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/blevesearch/bleve/pull/692#issuecomment-353793860>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASMnwxBmyvUzTCwJ1RygWMU_9HDkRZ0ks5tDoBqgaJpZM4RLTy3>
> .
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/blevesearch/bleve/pull/692#issuecomment-353794608>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAB3sMLv4BDIOBNk8LnyovXkV6iML3J-ks5tDoSngaJpZM4RLTy3>.

mschoch commented:
@steveyen here is one interesting thing, look at the keys:

```
bolt seekToKey ([116 1 0 52 255 57 57 54]) < prevKey ([116 1 0 53 255 48]), now at: [116 1 0 52 255 57 57 54]
```

116 = t (term freq row)
1 0 (field number 1)
52 255 (term '4')
57 57 54 (doc id '996')
and the prevKey was

116
1 0
53 255 (term '5')
48 (doc id '0')

But, this is a "term field reader", it is created to work on a single term, but the term itself has changed, not just the doc id (the things it should be nexting/advancing over).  I suspect this might be a upside_down instance reuse issue, not an actual seeking backwards issue, but still looking into exactly how it might happen.

mschoch commented:
Ah, OK, I think I understand now, and also why it only affects upside_down and not scorch.

In upside_down, a term field reader for '4' will actually go past the end when its done, pointing to the first key of the next term doc '0' of term '5' in that case.  Presumably it correctly signaled "end" in the higher-up code by comparing keys, but the underlying boltdb iterator was left pointing at whatever happened to be the next key.  We then allow advance to be called on it in the higher-up code and end up taking it backwards.

In scorch, you never end up with something accidentally pointing into the next term.  That said, I'm still not confident we handle this correctly, perhaps just slightly better.

What we need to ensure is that when the term field reader signals "end", that gets propagated correctly and we don't ever ask it to go backwards.

mschoch commented:
Confirming on the boltdb side, if you add && i.valid to the backwards check it never fires, indicating that in all cases where we actually seek backwards, it was because we gone off the end of the desired prefix/range.

mschoch commented:
So, in this particular case it looks like the reason that can happen is that on this line:

https://github.com/blevesearch/bleve/blob/scorch/search/searcher/search_phrase.go#L165-L166

We see that the phrase searcher advances its children for the subsequent iteration.  This means it can return a hit, and move its underlying searchers (and thus iterators) off the end.

The higher-level conjunction checks for nil, but that doesn't catch anything because it did return a non-nil hit.  So then it tells it to Advance, and this ends up asking searchers (and thus iterators) to advance to some key, even though they had reached the end.  This then shows up as a backwards movement.
